### PR TITLE
[closes #28] Add support for detectors

### LINKED ISSUE #28: Add support for assumptions
Issues are sometimes connected with specific environment and it would be nice to support this usecase.

e.g.
- Operation system
- JDK

### PR DESCRIPTION
**Pull request by basovnik**
This change is [Reviewable](https://reviewable.io/reviews/arquillian/arquillian-governor/29)

### CODE REVIEW

File: api/src/main/java/org/arquillian/extension/governor/api/detector/DetectorProcessor.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ */
+public class DetectorProcessor {
+
+    private static final Logger logger = Logger.getLogger(DetectorProcessor.class.getName());
+
+    public boolean process(Annotation annotation) {
+        boolean detected = true;
+        try {
+            final Detector detector = RefletionUtils.getAnnotationProperty(annotation, "detector", Detector.class);
```
Comment (smiklosovic) [inline api/src/main/java/org/arquillian/extension/governor/api/detector/DetectorProcessor.java]: RreflectionUtils

File: api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+package org.arquillian.extension.governor.api.detector;
+
+/**
+ * Interface {@code Strategy} can define merging strategy for objects of type {@link Detectable}.
+ *
+ * @author <a href="mailto:mbasovni@redhat.com">Martin Basovnik</a>
+ *
+ */
+public interface Strategy {
```
Comment (smiklosovic) [inline api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java]: DeciderStrategy

File: api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    boolean resolve();
+
+    class And extends BaseStrategy {
```
Comment (smiklosovic) [inline api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java]: is this public by default? i know that methods are but not sure abou classes

File: api/src/main/java/org/arquillian/extension/governor/utils/RefletionUtils.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package org.arquillian.extension.governor.utils;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public final class RefletionUtils {
```
Comment (smiklosovic) [inline api/src/main/java/org/arquillian/extension/governor/utils/RefletionUtils.java]: licence missing

**Comment by smiklosovic** [comment]
perfect work, done as we discussed, almost mergeable

**Comment by smiklosovic** [comment]
it is 2016 already, update the year in licences where applicable, could you add maven livence plugin to have this checked automatically in compile time please?

File: api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java (hunk 3 above)
Comment (basovnik) [inline api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java]: No, it is not public by deafault. I tried it:
```
[ERROR] /home/mbasovni/repos/git/github/arquillian/arquillian-governor/jira/src/test/java/org/arquillian/extension/governor/jira/ITJiraDetectorTest.java:[22,54] org.arquillian.extension.governor.api.detector.DeciderStrategy is not public in org.arquillian.extension.governor.api.detector; cannot be accessed from outside package
```

File: api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java (hunk 2 above)
Comment (basovnik) [inline api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java]: ok

File: api/src/main/java/org/arquillian/extension/governor/api/detector/DetectorProcessor.java (hunk 1 above)
Comment (basovnik) [inline api/src/main/java/org/arquillian/extension/governor/api/detector/DetectorProcessor.java]: ok, typo :)

File: api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java (hunk 3 above)
Comment (smiklosovic) [inline api/src/main/java/org/arquillian/extension/governor/api/detector/Strategy.java]: so please be sure to make it public then, thank you

**Comment by smiklosovic** [comment]
@lordofthejars @iocanel FYI totally custom deciders merged to Cube few days ago are on the way ... any hints for improvement?

**Comment by lordofthejars** [comment]
Ah right we had a plan to add these features in Governor. My question and I am talking without knowing exactly all the details. Why decider is tight to JIRA  or GitHub and not as a global annotation that can be used for any method being annotated with JIRA or not.

In case of Cube we need to decide what to execute but there are no issue (basically because are functional test/demos), so there is no JIRA or GitHub issue at all

**Comment by iocanel** [comment]
Also, if I am not mistaken those decider can only decide on test, while in the cube case we also need to be able to skip the entire class. In the long run we could replace the custom implementation in cube an use something powered by governor etc (the only blocker is being able to skip the entire class before even the arquillian descriptor gets processed).

**Comment by basovnik** [comment]
@lordofthejars, we intend to add support for multiple governor (e.g. `@Jira`) annotations for one method (#22). Our team needs it. Stefan already started with implementation. Then we would have to somehow map decider annotation with specific governor annotation and it would be complicated. So we decided to nest decider into the governor annotation as attribute.

@iocanel Support for class-level annotations is already implemented (#23).

@smiklosovic I added some default checkstyle configuration as you wished. It is not connected with this issue so I created another one (#30).

**Comment by iocanel** [comment]
@basovnik: In #23 you mention that 'arquillian cannot skip classes'. Has this changed? In our use case just skipping all test methods of a class does not work. We need to skip the class.

**Comment by basovnik** [comment]
@iocanel I guess your issue is connected with `@Deployment` method. I do not know how to easily fix your issue. Read this: https://github.com/arquillian/arquillian-governor/issues/23#issuecomment-215075667.

If you find a solution we can create another issue for that. What do you think?

**Comment by iocanel** [comment]
Yes, part of the issue is connected with @Deployment. In some other cases I think its even wider than it.

**Comment by smiklosovic** [comment]
@iocanel yes as far as I know this this holds, however, you could effectively skip whole class by skipping every method but I am not sure if such skipping is enough.

I was also thinking about making deciders standalone and not being tight to any issue tracker.

It would just be needed to take this API and make another Governor, completely unaware of issues logic, I can imaging it would be quite straight forward.

@basovnik we will merge this as is and we can elaborate on making them standalone in the next run. You did good thing you have put it into api so we are free to reuse it and implement it in a standalone way.

**Comment by basovnik** [comment]
**Ready to merge.**

_Note: Default java code style in IntelliJ IDEA can reformat code to proper style (checked by maven plugin) automatically._

**Comment by smiklosovic** [comment]
i can not merge it says there are merge conflicts

**Comment by smiklosovic** [comment]
you are just one commit behind master https://github.com/arquillian/arquillian-governor/commit/e71f621e0f0c0e3f0213e5c08f99b5b668c673e3

**Comment by basovnik** [comment]
Fixed.

**Comment by basovnik** [comment]
Thank you ;)

**Comment by smiklosovic** [comment]
thank you! :)

**Pull request merged**